test/library/draft/DistributedList/DistributedList.chpl

```diff
@@ -0,0 +1,763 @@
+                for locBlockIdx in 0..numBlockLayers {
+                    const blockStartingIdx = globalIndex(locIdx, locBlockIdx, 0),
+                          blockRange = blockStartingIdx..#blockSize;
+
+                    const globalIdxRange =
+                        if r.contains(blockRange) then blockRange
+                        else if r.contains(blockRange.first) then blockRange.first..r.last
+                        else if r.contains(blockRange.last) then r.first..blockRange.last
+                        else if blockRange.contains(r) then r
+                        else 0..<0;
+
+                    startsAndRanges[locIdx][locBlockIdx] = (blockStartingIdx, globalIdxRange);
+                }
+            }
+
+            return startsAndRanges;
+        }
+
+        pragma "no doc"
+        inline proc _destroy(ref item: eltType) {
+            chpl__autoDestroy(item);
+        }
+    }
+
+    // a list of blocks stored on a single locale
+    // as elements are added, the number of blocks can be expanded to accommodate them
+
+    //  within this data structure:
+    //  - 'block_idx' refers to the zero-based index of a block
+    //  - 'elt_idx' refers to the zero-based index of an element within a single block
+    record blockList {
+        type eltType;
+        param blockSize : int;  // the number of elements per block. Defaults to 16
+
+        var blocks: _ddata(_ddata(eltType)) = nil;
+
+        var numBlocks: int;     // the total number of blocks currently allocated
+        var numFilled: int;     // the total number of occupied elements in this blockList
+                                //  (the caller is expected to update 'numFilled' manually
+                                //   when modifying the blockList except when using the
+                                //   'appendBlock' method)
+
+        proc init(type eltType, param blockSize: int) {
+            this.eltType = eltType;
+            this.blockSize = blockSize;
+            this.numBlocks = DefaultNumBlocksPerLocale;
+            this.numFilled = 0;
+
+            this.complete();
+            on this {
+                this.blocks = this._makeBlockArray(this.numBlocks);
+                for i in 0..<this.numBlocks do this.blocks[i] = this._makeBlock();
+            }
+        }
+
+        // replace the current blocks with n new blocks
+        //  n defaults to twice the current number of blocks
+        proc acquireBlocks(n: int = this.numBlocks * 2) {
+            on this {
+                // allocate new blocks
+                assert(n >= this.numBlocks);
+                var blocks_new = this._makeBlockArray(n);
+
+                // transfer the old data into the new blocks
+                //  and free the old blocks
+                for b in 0..#this.numBlocks do
+                    if this.blocks[b] == nil
+                        then blocks_new[b] = this._makeBlock();
+                        else blocks_new[b] = this.blocks[b];
```

Comment:
Does this free the old block?  The comment says it does but I don't see an action that would.  Is the comment correct?

Reply:
I think that was an old comment from when I was doing this a different way. I'll remove it.